flink-ml-tests/src/test/java/org/apache/flink/test/iteration/BoundedIterationWithBroadcastITCase.java

```diff
@@ -0,0 +1,152 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.iteration;
+
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.iteration.DataStreamList;
+import org.apache.flink.iteration.IterationBody;
+import org.apache.flink.iteration.IterationBodyResult;
+import org.apache.flink.iteration.IterationConfig;
+import org.apache.flink.iteration.IterationConfig.OperatorLifeCycle;
+import org.apache.flink.iteration.Iterations;
+import org.apache.flink.iteration.ReplayableDataStreamList;
+import org.apache.flink.ml.common.broadcast.BroadcastUtils;
+import org.apache.flink.ml.common.iteration.TerminateOnMaxIter;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.testutils.junit.SharedObjects;
+import org.apache.flink.testutils.junit.SharedReference;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static org.apache.flink.test.iteration.UnboundedStreamIterationITCase.createMiniClusterConfiguration;
+import static org.junit.Assert.assertEquals;
+
+/** Tests using {@link BroadcastUtils#withBroadcastStream} in iterations. */
+public class BoundedIterationWithBroadcastITCase extends TestLogger {
+
+    @Rule public final SharedObjects sharedObjects = SharedObjects.create();
+
+    private MiniCluster miniCluster;
+
+    private SharedReference<BlockingQueue<Long>> result;
+
+    @Before
+    public void setup() throws Exception {
+        miniCluster = new MiniCluster(createMiniClusterConfiguration(1, 1));
+        miniCluster.start();
+
+        result = sharedObjects.add(new LinkedBlockingQueue<>());
+    }
+
+    @After
+    public void teardown() throws Exception {
+        if (miniCluster != null) {
+            miniCluster.close();
+        }
+    }
+
+    @Test
+    public void testIterationWithBroadcastJobGraph() throws Exception {
```

Comment:
Can you help explain why `ListStateDescriptor` is not initialized in this test but initialized in other tests?

Reply:
The `getListState` in `DefaultOperatorStateBackend` does the initialization here [1] while `ProxyOperatorStateBackend` does not.

Reply:
Suppose that `ProxyOperatorStateBackend#wrappedBackend` is a `DefaultOperatorStateBackend`. Then `ProxyOperatorStateBackend#getUnionListState` invokes `DefaultOperatorStateBackend#getUnionListState`, which should in turn invoke `DefaultOperatorStateBackend#getListState` and initialize the state descriptor, right?

Reply:
Before calling `DefaultOperatorStateBackend#getListState`, we need to call `stateDescriptor.getElementSerializer()`. If it is null, it throws NPE here.

Reply:
Before calling `DefaultOperatorStateBackend#getListState`, `ProxyOperatorStateBackend` will call `getElementSerializer` first, this may lead to NPE.